test/e2e/addon/kubestatemetrics.go

```diff
@@ -0,0 +1,186 @@
+package addon
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/service/eks"
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	clientgo "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	ik8s "github.com/aws/eks-hybrid/internal/kubernetes"
+	"github.com/aws/eks-hybrid/internal/retry"
+	"github.com/aws/eks-hybrid/test/e2e/kubernetes"
+)
+
+const (
+	kubeStateMetricsName        = "kube-state-metrics"
+	kubeStateMetricsNamespace   = "kube-state-metrics"
+	kubeStateMetricsWaitTimeout = 2 * time.Minute
+)
+
+// KubeStateMetricsTest tests the kube-state-metrics addon
+type KubeStateMetricsTest struct {
+	Cluster   string
+	addon     *Addon
+	K8S       clientgo.Interface
+	EKSClient *eks.Client
+	K8SConfig *rest.Config
+	Logger    logr.Logger
+}
+
+// Create installs the kube-state-metrics addon
+func (k *KubeStateMetricsTest) Create(ctx context.Context) error {
+	k.addon = &Addon{
+		Cluster:   k.Cluster,
+		Namespace: kubeStateMetricsNamespace,
+		Name:      kubeStateMetricsName,
+	}
+
+	if err := k.addon.CreateAndWaitForActive(ctx, k.EKSClient, k.K8S, k.Logger); err != nil {
+		return err
+	}
+
+	if _, err := ik8s.GetAndWait(ctx, kubeStateMetricsWaitTimeout, k.K8S.AppsV1().Deployments(k.addon.Namespace), k.addon.Name, func(d *appsv1.Deployment) bool {
+		return d.Status.Replicas == d.Status.ReadyReplicas
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Validate checks if kube-state-metrics is providing metrics
+func (k *KubeStateMetricsTest) Validate(ctx context.Context) error {
+	k.Logger.Info("Checking if kube-state-metrics is providing metrics")
+
+	// Find the service for kube-state-metrics
+	svc, err := ik8s.GetRetry(ctx, k.K8S.CoreV1().Services(kubeStateMetricsNamespace), kubeStateMetricsName)
+	if err != nil {
+		return fmt.Errorf("failed to get service: %v", err)
+	}
+
+	if len(svc.Spec.Ports) == 0 {
+		return fmt.Errorf("no ports found for service")
+	}
+
+	// port is typically 8080
+	var metricsPort int32
+	for _, port := range svc.Spec.Ports {
+		if port.Name == "http" || port.Name == "metrics" {
+			metricsPort = port.Port
+			break
+		}
+	}
+
+	if metricsPort == 0 {
+		metricsPort = svc.Spec.Ports[0].Port // Fallback to first port if named ports not found
+	}
+
+	k.Logger.Info("Found service", "name", svc.Name, "port", metricsPort)
+
+	// Construct metrics endpoint URL
+	metricsEndpoint := fmt.Sprintf("%s/api/v1/namespaces/%s/services/%s:%d/proxy/metrics",
```

Comment:
i guess there is no api method for this which is why we have to hand code the http stuff?

Reply:
yeah